[AIRAVATA-2662] Adding Migrator class to create everyone group in all domains

tilaks26 opened the pull request:
One-time migration to create and populate the “everyone” group for all existing domains and share all the existing application deployments with the group.

machristie commented on modules/sharing-registry/sharing-registry-server/src/main/java/org/apache/airavata/sharing/registry/server/SharingRegistryServerHandler.java:
```diff
@@ -277,7 +277,9 @@ public String createGroup(UserGroup group) throws SharingRegistryException, TExc
             (new UserGroupRepository()).create(group);
 
-            addUsersToGroup(group.domainId, Arrays.asList(group.ownerId), group.groupId);
+            if(group.ownerId != null)
```
Sneha, instead of having no owner, let's make the owner an argument to the migration script like you did in the other pull request.

tilaks26 replied:
Sure.

machristie commented:
@tilaks26 , can you also change the sharing registry so that whenever a new user is created it is added to the everyone group?  The migration script will add all existing users to the everyone group but we need to also add any new users that get created after the migration script runs to the everyone group.

tilaks26 commented:
@machristie Should that be a part of the createUser() method in the SharingRegistryServerHandler or do you want me to make a new method for adding new users to the everyone group?

tilaks26 commented:
@machristie , I've made the requested changes.

machristie commented on airavata-api/airavata-api-server/src/main/java/org/apache/airavata/api/server/handler/AiravataServerHandler.java:
```diff
@@ -338,6 +338,20 @@ public String addGateway(AuthzToken authzToken, Gateway gateway) throws InvalidR
+            userGroup.setGroupCardinality(GroupCardinality.MULTI_USER);
+            userGroup.setCreatedTime(System.currentTimeMillis());
+            userGroup.setUpdatedTime(System.currentTimeMillis());
+            userGroup.setOwnerId(authzToken.getClaimsMap().get(Constants.USER_NAME));
```
OwnerId should be in the form "username@domainId"

tilaks26 replied:
Resolved.

tilaks26 commented:
Changes to the AiravataDataMigrator.java class:
- Removed the EveryoneGroupMigrator.java class and included the changes in AiravataDataMigrator.java class
- Modified the pom.xml to include the security services stubs as well
- Retrieving groupOwner and ApplicationDeploymentOwner from the KeyCloakSecurityManager.java class
- Added a new PermissionType - EXEC for Application Deployments

Changes to the AiravataServerHandler.java class:
- Added a new PermissionType - EXEC for methods - initSharingRegistry(), addGateway(), shareResourceWithUsers() and revokeSharingOfResourceFromUsers()
- Modified getAccessibleApplicationDeployments() to first retrieve all the Group Resource Profiles accessible by the user in the gateway and extracting all the Compute Resource IDs from the list of Group Resource Profiles which is then passed to the getAccessibleApplicationDeployments() of the RegsitryServerHandler.java class

Changes to the registry-api.thrift, RegistryServerHandler.java class and ApplicationDeployment.java interface:
- Modified getAccessibleApplicationDeployments() to accept the accessible Compute Resource IDs
- Regenerated the Thrift files for Registry

Changes to the AppDeploymentResource.java class:
- Modified getAll() to also filter by the accessible Compute Resource IDs

tilaks26 commented:
This Pull Request contains the fixes and suggestions for [AIRAVATA-2690] ticket as well.

machristie commented on airavata-api/airavata-api-server/src/main/java/org/apache/airavata/api/server/handler/AiravataServerHandler.java:
```diff
@@ -4926,9 +4965,12 @@ public boolean shareResourceWithUsers(AuthzToken authzToken, String resourceId,
+                else if(userPermission.getValue().equals(ResourcePermissionType.READ))
                     sharingClient.shareEntityWithUsers(gatewayId, resourceId,
                             Arrays.asList(userPermission.getKey()), authzToken.getClaimsMap().get(Constants.GATEWAY_ID) + ":" + "READ", true);
+                else
```
Can you check for the EXEC permission here and throw an exception in the else block? And on line 5004 as well?

machristie commented on modules/sharing-registry/sharing-data-migrator/src/main/java/org/apache/airavata/sharing/registry/migrator/airavata/AiravataDataMigrator.java:
```diff
@@ -205,17 +218,56 @@ public static void main(String[] args) throws SQLException, ClassNotFoundExcepti
+        //Creating the everyone group
+        List<Domain> domainList = sharingClient.getDomains(-1, 0);
+        for (Domain domain : domainList) {
+            CredentialStoreService.Client credentialStoreServiceClient = getCredentialStoreServiceClient();
```
You can move this out of the loop, no need to create one each time.

tilaks26 replied:
credentialStoreServiceClient is now initialized outside the loop.

machristie commented on modules/sharing-registry/sharing-data-migrator/src/main/java/org/apache/airavata/sharing/registry/migrator/airavata/AiravataDataMigrator.java:
```diff
@@ -205,17 +218,56 @@ public static void main(String[] args) throws SQLException, ClassNotFoundExcepti
         rs = statement.executeQuery(query);
         while(rs.next()){
             try {
+                CredentialStoreService.Client credentialStoreServiceClient = getCredentialStoreServiceClient();
```
You probably don't need this here if you moved the credentialStoreServiceClient out of the loop above.

tilaks26 replied:
Done.

machristie commented on modules/sharing-registry/sharing-registry-server/pom.xml:
```diff
@@ -47,6 +47,11 @@
             <artifactId>airavata-sharing-registry-stubs</artifactId>
             <version>${project.parent.version}</version>
         </dependency>
+        <dependency>
```
Is this needed?

tilaks26 replied:
No, Marcus. Will remove it.

machristie commented on modules/sharing-registry/sharing-data-migrator/pom.xml:
```diff
@@ -59,6 +59,16 @@
+            <artifactId>airavata-credential-store-stubs</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
```
Do you need the services-security dependency?

tilaks26 replied:
You're right. This is not needed. I added this earlier because I was using the KeyCloakSecurityManager instance in my code.

machristie commented:
@tilaks26 this looks good, I've reviewed the changes.  Two other things:

Looks like there is a build error, the error message is
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.6.1:compile (default-compile) on project airavata-registry-core: Compilation failure: Compilation failure: 
[ERROR] /home/travis/build/apache/airavata/modules/registry/registry-core/src/main/java/org/apache/airavata/registry/core/app/catalog/impl/ApplicationDeploymentImpl.java:[38,8] org.apache.airavata.registry.core.app.catalog.impl.ApplicationDeploymentImpl is not abstract and does not override abstract method getAccessibleApplicationDeployements(java.lang.String,java.util.List<java.lang.String>,java.util.List<java.lang.String>) in org.apache.airavata.registry.cpi.ApplicationDeployment
[ERROR] /home/travis/build/apache/airavata/modules/registry/registry-core/src/main/java/org/apache/airavata/registry/core/app/catalog/impl/ApplicationDeploymentImpl.java:[398,5] method does not override or implement a method from a supertype
[ERROR] /home/travis/build/apache/airavata/modules/registry/registry-core/src/main/java/org/apache/airavata/registry/core/app/catalog/impl/ApplicationDeploymentImpl.java:[404,21] cannot find symbol
[ERROR]   symbol:   method setAccessibleAppDeploymentIds(java.util.List<java.lang.String>)
[ERROR]   location: variable resource of type org.apache.airavata.registry.core.app.catalog.resources.AppDeploymentResource
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :airavata-registry-core
```

Also, can you update getAccessibleAppModules in the AiravataServerHandler, similar to what you did with getAccessibleApplicationDeployments?

machristie commented on modules/sharing-registry/sharing-data-migrator/pom.xml:
```diff
@@ -94,5 +94,10 @@
             <artifactId>slf4j-api</artifactId>
             <version>1.7.12</version>
         </dependency>
+        <dependency>
```
Thanks @smarru , I'll check if the migrator is included in the distribution.

machristie replied:
I verified that the migrator isn't included in the distribution.
